File: 2021/day06-2.ts
type FishGroup = {
  day: number;
  size: number;
};

function solution(content: string) {
  return Array.from({ length: 256 })
    .reduce((acc: FishGroup[]) => play(acc), parse(content))
    .reduce((acc, { size }) => acc + size, 0);
}

function play(groups: FishGroup[]): FishGroup[] {
  return clean(nextDay(generate(groups)));
}

function clean(groups: FishGroup[]): FishGroup[] {
  return groups
    .filter(({ day }) => day !== 6)
    .concat([
      {
        day: 6,
        size: groups
          .filter(({ day }) => day === 6)
          .map(({ size }) => size)
          .reduce((acc, cur) => acc + cur, 0),
      },
    ])
    .filter(({ size }) => size != 0);
}

function nextDay(groups: FishGroup[]): FishGroup[] {
  return groups.map((g) => ({
    day: g.day === 0 ? 6 : g.day - 1,
    size: g.size,
  }));
}

function generate(groups: FishGroup[]): FishGroup[] {
  return groups.concat(
    groups.filter(({ day }) => day === 0).map(({ size }) => ({ day: 9, size }))
  );
}

function parse(content: string): FishGroup[] {
  const data = content.split(",").map(Number);
  return Array.from(new Set(data)).map((day) => ({
    day,
    size: data.filter((d) => d === day).length,
  }));
}

Deno.readTextFile("day06.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);

File: 2021/day06_2.py
# pylint: disable=missing-module-docstring, missing-function-docstring


from typing import Any

type FishGroup = tuple[int, int]


def solution(content: str) -> Any:
    groups = parse(content)
    for _ in range(256):
        groups = play(groups)
    return sum(size for _, size in groups)


def play(groups: list[FishGroup]) -> list[FishGroup]:
    return clean(next_day(generate(groups)))


def clean(groups: list[FishGroup]) -> list[FishGroup]:
    copy = groups.copy()
    copy = [(day, size) for day, size in copy if day != 6]
    copy.append((6, sum(size for day, size in groups if day == 6)))
    copy = [(day, size) for day, size in copy if size != 0]
    return copy


def next_day(groups: list[FishGroup]) -> list[FishGroup]:
    return [(6 if day == 0 else (day - 1), size) for day, size in groups]


def generate(groups: list[FishGroup]) -> list[FishGroup]:
    copy = groups.copy()
    copy.extend((9, size) for day, size in groups if day == 0)
    return copy


def parse(content: str) -> list[FishGroup]:
    data = [int(x) for x in content.split(",")]
    return [(day, len([d for d in data if d == day])) for day in set(data)]


def main() -> None:
    with open("day06.dat", "r", encoding="utf8") as file:
        print(solution(file.read().strip()))


if __name__ == "__main__":
    main()
